lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -997,6 +990,18 @@ SerializedModuleLoaderBase::loadModule(SourceLoc importLoc,
   } else {
     M->setFailedToLoad();
   }
+
+  if (dependencyTracker && file) {
+    auto DepPath = file->getFilename();
+    // Don't record cached artifacts as dependencies.
+    if (!isCached(DepPath)) {
+      if (M->hasIncrementalInfo()) {
+        dependencyTracker->addIncrementalDependency(DepPath);
+      } else {
+        dependencyTracker->addDependency(DepPath, /*isSystem=*/false);
+      }
+    }
+  }
```

Comment:
my own style would be to factor these lines out into a function, "trackDependency" or some such.
Also, it would be helpful to add a "why" to the comment on 996. (Why not record cached artifacts?)